shared/crypto/bls.go

```diff
@@ -0,0 +1,135 @@
+	X := suite.G2().Point().Mul(x, nil)
+	return x, X
+}
+
+// Sign creates a BLS signature S = x * H(m) on a message m using the private
+// key x. The signature S is a point on curve G1.
+func Sign(suite pairing.Suite, x kyber.Scalar, msg []byte) ([]byte, error) {
+	hashable, ok := suite.G1().Point().(hashablePoint)
+	if !ok {
+		return nil, errors.New("point needs to implement hashablePoint")
+	}
+	HM := hashable.Hash(msg)
+	xHM := HM.Mul(x, HM)
+
+	s, err := xHM.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// AggregateSignatures combines signatures created using the Sign function
+func AggregateSignatures(suite pairing.Suite, sigs ...[]byte) ([]byte, error) {
+	sig := suite.G1().Point()
+	for _, sigBytes := range sigs {
+		sigToAdd := suite.G1().Point()
+		if err := sigToAdd.UnmarshalBinary(sigBytes); err != nil {
+			return nil, err
+		}
+		sig.Add(sig, sigToAdd)
+	}
+	return sig.MarshalBinary()
+}
+
+// AggregatePublicKeys takes a slice of public G2 points and returns
+// the sum of those points. This is used to verify multisignatures.
+func AggregatePublicKeys(suite pairing.Suite, Xs ...kyber.Point) kyber.Point {
+	aggregated := suite.G2().Point()
+	for _, X := range Xs {
+		aggregated.Add(aggregated, X)
+	}
+	return aggregated
+}
+
+// BatchVerify verifies a large number of publicKey/msg pairings with a single aggregated signature.
+// Since aggregation is generally much faster than verification, this can be a speed enhancement.
+// Benchmarks show a roughly 50% performance increase over individual signature verification
+// Every msg must be unique or there is the possibility to accept an invalid signature
+// see: https://crypto.stackexchange.com/questions/56288/is-bls-signature-scheme-strongly-unforgeable/56290
+// for a description of why each message must be unique.
+func BatchVerify(suite pairing.Suite, publics []kyber.Point, msgs [][]byte, sig []byte) error {
+	if !distinct(msgs) {
+		return fmt.Errorf("bls: error, messages must be distinct")
+	}
+
+	s := suite.G1().Point()
+	if err := s.UnmarshalBinary(sig); err != nil {
+		return err
+	}
+
+	var aggregatedLeft kyber.Point
+	for i := range msgs {
+		hashable, ok := suite.G1().Point().(hashablePoint)
+		if !ok {
+			return errors.New("bls: point needs to implement hashablePoint")
+		}
+		hm := hashable.Hash(msgs[i])
+		pair := suite.Pair(hm, publics[i])
+
+		if i == 0 {
+			aggregatedLeft = pair
+		} else {
+			aggregatedLeft.Add(aggregatedLeft, pair)
+		}
+	}
+
+	right := suite.Pair(s, suite.G2().Point().Base())
+	if !aggregatedLeft.Equal(right) {
+		return errors.New("bls: invalid signature")
+	}
+	return nil
+}
+
+// Verify checks the given BLS signature S on the message m using the public
+// key X by verifying that the equality e(H(m), X) == e(H(m), x*B2) ==
+// e(x*H(m), B2) == e(S, B2) holds where e is the pairing operation and B2 is
+// the base point from curve G2.
+func Verify(suite pairing.Suite, X kyber.Point, msg, sig []byte) error {
```

Comment:
NIT: rename vars to match comment?

Reply:
"Verify" is what they call this in the IETF spec for BLS so I want to keep that for that reason, so I'll change the comment

Reply:
https://www.ietf.org/id/draft-irtf-cfrg-bls-signature-05.html